### PR TITLE
feat: add Filterable MultiSelect component

### LINKED ISSUE #386: [Component] Combobox, MultiSelect, etc. variants
This issue is for building out the following table of components, using work from the [`downshift`](https://github.com/paypal/downshift) library.

| Component | Selection mode | Type | Filterable | Status |
|-------------|----------------|-------|---------|:----:|
| Dropdown | Single | Default | No | ✅|
| Inline Dropdown | Single | Inline | No | ✅ |
| MultiSelect | Multi | Default | No | ✅ |
| Inline MultiSelect | Multi | Inline | No | ✅ |
| Combobox | Single | Default | Yes | ✅  |
| MultiSelect Combobox | Multi | Default | Yes | ✅ |

### PR DESCRIPTION
**Pull request by joshblack**
Closes #386

Adds `MultiSelect.Filterable` as a component to the codebase.

#### Changelog

**New**

- `src/components/FilterableMultiSelect.js`

**Changed**

- Pull out tools for ComboBoxes into `src/components/ComboBox/tools`
- Update ListBox `test-helpers`
- Update `MultiSelect` story to include filterable story
- Extract out `sortingProps` for `MultiSelect`

**Removed**

### CODE REVIEW

**Comment by marijohannessen** [comment]
For some reason the checkboxes are acting a bit weird in Firefox:

screen shot 2018-01-19 at 1 49 56 pm
screen shot 2018-01-19 at 1 50 00 pm

When I re-open the dropdown they look normal, but selecting an option makes them shift a bit. Does not happen in any other browsers 🤔

File: src/components/MultiSelect/MultiSelect-story.js [hunk 1]
```diff
@@ -81,8 +81,58 @@ storiesOf('MultiSelect', module)
...
+    `
+      Filterable version of our MultiSelect component
+    `,
+    () => (
+      <div style={{ width: 300 }}>
+        <MultiSelect.Filterable
+          items={items}
+          itemToString={item => (item ? item.text : '')}
+          onChange={action('onChange')}
+          placeholder="Placeholder..."
```
Comment (marijohannessen) [inline src/components/MultiSelect/MultiSelect-story.js]: Very minor thing - but do you want to check with @bsonefeld what this placeholder should say? We try and keep the design kit and the coded components in sync as much as possible, and I _think_ it just says `Filter` in the design kit but I might be wrong!

Comment (joshblack) [reply src/components/MultiSelect/MultiSelect-story.js]: Cool! Makes sense to me. Should I just grab the placeholder from the dropdown stuff?

Comment (marijohannessen) [reply src/components/MultiSelect/MultiSelect-story.js]: Yeah, either just grab it from the designs or ask one of the designers directly! 🙂

Comment (bsonefeld) [reply src/components/MultiSelect/MultiSelect-story.js]: Yep, @marijohannessen 's right, the placeholder should say `Filter`. 

Ref: 
screen shot 2018-01-19 at 2 19 46 pm

Comment (joshblack) [reply src/components/MultiSelect/MultiSelect-story.js]: Updated!

**Comment by joshblack** [comment]
Pushed the updates for content, unfortunately can't replicate the CSS issue @marijohannessen in 52.5.3. Have any suggestions to help isolate it?

screen shot 2018-01-22 at 10 16 17 am

**Comment by marijohannessen** [comment]
Updating to version `8.7.1` of `carbon-components` fixed the Checkbox issue for me! 🎉 👍

**Pull request merged**